app/controllers/posts_controller.rb

```diff
@@ -2,16 +2,22 @@ class PostsController < ApplicationController
   def index
     @user = User.find(params[:user_id])
     @posts = @user.posts.includes(:comments)
+  rescue StandardError
+    redirect_to root_path, error: 'The page you\'re trying to access does not exist!'
   end
 
   def show
     @current_user = current_user
     @user = User.find(params[:user_id])
     @post = Post.find(params[:id])
-    @user_liked_post = current_user.likes.where(post: @post).any?
+    @user_liked_post = current_user.likes.where(post: @post).any? if current_user
+  rescue StandardError
+    flash[:error] = 'The page you\'re trying to access does not exist!'
+    redirect_to root_path
   end
 
```

Comment:
- [ ]  You did great job to implement the authorization using CanCan, but I would suggest you to consider using `load_and_authorize_resource` in your post/comment controller for it will automatically loads the appropriate model instance and authorizes it based on the specified permissions.  For [more ](https://github.com/CanCanCommunity/cancancan)